### PR TITLE
[SYCL][GRAPH] Add graph E2E tests which use free function extension

### PR DESCRIPTION
**Pull request by konradkusiak97**
Dynamic parameters in SYCL Graphs should be used for now with free_function_kernels extension since then the order of kernel arguments is well defined. 

The tests are a copy of few selected ones from `Graph/Update` and modified such that they use the extension. They live in `Graph/Update/FreeFunctionKernels`.

 Since the free function kernels are not yet implemented on CUDA, we should keep for now both copies of the tests (with and without free functions).

### CODE REVIEW

**Review by EwanC — COMMENTED** [review]
Couple of comments:

* Add an entry to this part of the spec https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc#712-interaction-with-other-extensions to say we support the free function extension (and add yourself to the contributors section)

* These tests all rely on a device having update support, however even without update support a device should still be able to use the free function extension in a static graph. Could we have a test for that? Probably only need 1 under `Inputs` which is then used in `Explicit` and `RecordReplay` variants.

File: sycl/test-e2e/Graph/Update/FreeFunctionKernels/update_before_finalize.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+  std::vector<int> HostDataA(N);
+  std::vector<int> HostDataB(N);
+
+  Queue.memset(PtrA, 0, N * sizeof(int)).wait();
+  Queue.memset(PtrB, 0, N * sizeof(int)).wait();
+
+  exp_ext::dynamic_parameter InputParam(Graph, PtrA);
+
+#ifndef __SYCL_DEVICE_ONLY__
```
[the remark this answers is not in the document]
Comment (konradkusiak97) [reply sycl/test-e2e/Graph/Update/FreeFunctionKernels/update_before_finalize.cpp:35]: I think they're required. I'm not exactly sure why but the tests don't compile without them. Similarly in other tests which use free function kernels they're present, e.g. [here](https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/KernelAndProgram/free_function_kernels.cpp)

File: sycl/test-e2e/Graph/Update/FreeFunctionKernels/update_before_finalize.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+int main() {
+  queue Queue{};
+  context ctxt{Queue.get_context()};
+
+  const size_t N = 1024;
+
+  exp_ext::command_graph Graph{ctxt, Queue.get_device()};
+
+  int *PtrA = malloc_device<int>(N, Queue);
+  int *PtrB = malloc_device<int>(N, Queue);
```
[the remark this answers is not in the document]
Comment (konradkusiak97) [reply sycl/test-e2e/Graph/Update/FreeFunctionKernels/update_before_finalize.cpp]: I fixed it in the new tests, I'll make a separate PR with fixing that in the other tests form `Update/`

File: sycl/test-e2e/Graph/Update/FreeFunctionKernels/update_before_finalize.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (EwanC) [reply sycl/test-e2e/Graph/Update/FreeFunctionKernels/update_before_finalize.cpp:33]: Interesting, that's good to know 👍

File: sycl/test-e2e/Graph/Inputs/free_function_kernels.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+  auto ExecGraph = Graph.finalize();
+
+  Queue.ext_oneapi_graph(ExecGraph).wait();
+
+  Queue.copy(PtrA, HostDataA.data(), Size).wait();
+  for (size_t i = 0; i < Size; i++) {
+    assert(HostDataA[i] == i);
+  }
+  sycl::free(PtrA, Queue);
```
Comment (EwanC) [inline sycl/test-e2e/Graph/Inputs/free_function_kernels.cpp:41]: The ` malloc_device<int>(Size, Queue);` isn't inside the `#ifdef` so the free should probably be outside it too

File: sycl/test-e2e/Graph/Inputs/free_function_kernels.cpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((exp_ext::single_task_kernel))
+void ff_0(int *ptr) {
+  for (size_t i{0}; i < Size; ++i) {
+    ptr[i] = i;
+  }
+}
+
+int main() {
+  queue Queue{};
+  context ctxt{Queue.get_context()};
```
Comment (EwanC) [inline sycl/test-e2e/Graph/Inputs/free_function_kernels.cpp:14]: `Ctxt` to match variable naming conventions (same in other files)

File: sycl/test-e2e/Graph/Inputs/free_function_kernels.cpp [hunk 5]
```diff
@@ -0,0 +1,44 @@
+// Tests compatibility with free function kernels extension
+
+#include "../graph_common.hpp"
+
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((exp_ext::single_task_kernel))
+void ff_0(int *ptr) {
```
Comment (EwanC) [inline sycl/test-e2e/Graph/Inputs/free_function_kernels.cpp:6]: `Ptr` to match variable naming conventions

File: sycl/test-e2e/Graph/Update/FreeFunctionKernels/free_function_kernels.hpp [hunk 6]
```diff
@@ -0,0 +1,54 @@
+#include "../../graph_common.hpp"
+#include "sycl/ext/oneapi/kernel_properties/properties.hpp"
+#include "sycl/kernel_bundle.hpp"
+#include <sycl/ext/oneapi/free_function_queries.hpp>
+
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((exp_ext::single_task_kernel))
+void ff_0(int *ptr) {
```
Comment (EwanC) [inline sycl/test-e2e/Graph/Update/FreeFunctionKernels/free_function_kernels.hpp:7]: Variables names in this file for parameters should start with an uppercase

File: sycl/test-e2e/Graph/Update/FreeFunctionKernels/free_function_kernels.hpp [hunk 7]
```diff
@@ -0,0 +1,54 @@
+#include "../../graph_common.hpp"
```
Comment (EwanC) [inline sycl/test-e2e/Graph/Update/FreeFunctionKernels/free_function_kernels.hpp:1]: `#pragma once` here too?

**Review by EwanC — APPROVED** [review]
LGTM, remaining comments are nitpicking

File: sycl/test-e2e/Graph/Inputs/free_function_kernels.cpp (hunk 3 above)
Comment (konradkusiak97) [reply sycl/test-e2e/Graph/Inputs/free_function_kernels.cpp]: Done.

**Comment by konradkusiak97** [comment]
@intel/llvm-gatekeepers could you merge this, please?

**Pull request merged**